iroh-net/src/magicsock/peer_map/endpoint.rs

```diff
@@ -199,68 +223,93 @@ impl Endpoint {
-                trace!(addr = %best_addr.addr, latency = ?best_addr.latency, "best_addr is set and valid, use best_addr only");
-                (Some(best_addr.addr), None, false)
+                // If we have a valid address we use it.
+                trace!(addr = %best_addr.addr, latency = ?best_addr.latency,
+                       "best_addr is set and valid, use best_addr only");
+                (Some(best_addr.addr), None)
             }
-            // we have an outdated address: use it, but use derp as well.
             best_addr::State::Outdated(best_addr) => {
-                trace!(addr = %best_addr.addr, latency = ?best_addr.latency, "best_addr is set but outdated, use best_addr and derp");
-                (Some(best_addr.addr), self.derp_url(), true)
+                // If the address is outdated we use it, but send via derp at the same time.
+                // We also send disco pings so that it will become valid again if it still
+                // works (i.e. we don't need to holepunch again).
+                trace!(addr = %best_addr.addr, latency = ?best_addr.latency,
+                       "best_addr is set but outdated, use best_addr and derp");
+                (Some(best_addr.addr), self.derp_url())
             }
-            // we have no best address: use a random candidate if available, and derp as backup.
             best_addr::State::Empty => {
+                // No direct connection has been used before.  If we know of any possible
+                // candidate addresses, randomly try to use one while also sending via derp
+                // at the same time.  If we use a candidate address also send disco ping
+                // message to it.
                 let addr = self
                     .direct_addr_state
                     .keys()
+                    .filter(|ipp| match ipp.ip() {
+                        IpAddr::V4(_) => true,
+                        IpAddr::V6(_) => have_ipv6,
+                    })
                     .choose_stable(&mut rand::thread_rng())
                     .map(|ipp| SocketAddr::from(*ipp));
                 trace!(udp_addr = ?addr, "best_addr is unset, use candidate addr and derp");
-                let should_ping = addr.is_some()
-                    || self
-                        .derp_url
-                        .as_ref()
-                        .map(|(_r, state)| state.needs_ping(now))
-                        .unwrap_or(false);
-                (addr, self.derp_url(), should_ping)
+                (addr, self.derp_url())
             }
         }
     }
 
-    /// Update our best_addr (if empty) with the candidate udp addr with the lowest latency.
+    /// Fixup best_addr from candidates.
+    ///
+    /// If somehow we end up in a state where we failed to set a best_addr, while we do have
+    /// valid candidates, this will chose a candidate and set best_addr again.  Most likely
+    /// this is a bug elsewhere though.
```

Comment:
all changes in this function are a re-write for readability, should not contain functional changes